csharp/src/Ice/LocatorInfo.cs

```diff
@@ -266,132 +308,208 @@ private static void TraceInvalid(Reference reference, Reference invalidReference
-            sb.Append("location = ");
-            sb.Append(indirectReference.LocationAsString);
-            wellKnown.Communicator.Logger.Trace(TraceLevels.LocatorCategory, sb.ToString());
+            if (_wellKnownProxyCache.TryRemove(
+                (identity, protocol),
+                out (TimeSpan _, EndpointList Endpoints, Location Location) entry))
+            {
+                if (entry.Endpoints.Count > 0)
+                {
+                    if (communicator.TraceLevels.Locator >= 2)
+                    {
+                        Trace("removed well-known proxy with endpoints from locator cache",
+                              identity,
+                              protocol,
+                              entry.Endpoints,
+                              entry.Location,
+                              communicator);
+                    }
+                }
+                else
+                {
+                    Debug.Assert(entry.Location.Count > 0);
+                    if (communicator.TraceLevels.Locator >= 2)
+                    {
+                        Trace("removed well-known proxy without endpoints from locator cache",
+                              identity,
+                              protocol,
+                              entry.Location,
+                              communicator);
+                    }
+
+                    ClearCache(entry.Location, protocol, communicator);
+                }
+            }
         }
 
-        private (Reference? Reference, bool Cached) GetResolvedLocationFromCache(Reference reference, TimeSpan ttl)
+        private void ClearCache(Location location, Protocol protocol, Communicator communicator)
         {
-            if (ttl == TimeSpan.Zero) // Locator cache disabled.
+            if (_locationCache.TryRemove(
+                (location, protocol),
+                out (TimeSpan _, EndpointList Endpoints, Location Location) entry))
             {
-                return (null, false);
+                if (communicator.TraceLevels.Locator >= 2)
+                {
+                    Trace("removed endpoints for location from locator cache",
+                          location,
+                          protocol,
+                          entry.Endpoints,
+                          entry.Location,
+                          communicator);
+                }
             }
+        }
 
-            if (_locationCache.TryGetValue(reference, out (TimeSpan InsertionTime, Reference Reference) entry))
+        private (EndpointList Endpoints, Location Location, bool Cached)
+        GetResolvedLocationFromCache(
+            Location location,
+            Protocol protocol,
+            TimeSpan ttl)
+        {
+            if (ttl != TimeSpan.Zero && _locationCache.TryGetValue(
+                (location, protocol),
+                out (TimeSpan InsertionTime,
+                    EndpointList Endpoints,
+                    Location Location) entry))
```

Comment:
Perhaps using a struct instead the tuple would improve readability a bit?

Reply:
It actually fits on a single line.